### PR TITLE
mef: reindex during entities reindexing

### PR DESCRIPTION
**Pull request by rerowep**
Co-Authored-by: Peter Weber <peter.weber@rero.ch>

### CODE REVIEW

**Review by BadrAly — APPROVED** [review]
`during` instead of `durin` in commit message

**Review by jma — CHANGES_REQUESTED** [review]
Please fix you commit message. Tests are missing.

File: rero_mef/agents/api.py [hunk 1]
```diff
@@ -72,31 +65,36 @@ def create_or_update_mef_viaf_record(self, dbcommit=False, reindex=False,
...
-                    viaf_record.pid)
-        if self.deleted:
-            mef_record, mef_action = self.delete_from_mef(
-                dbcommit=dbcommit,
-                reindex=reindex
+        from .viaf.api import AgentViafRecord
+        if not isinstance(self, (AgentViafRecord, AgentMefRecord)):
+            AgentViafRecord.update_indexes()
+            viaf_record, got_online = AgentViafRecord.get_viaf_by_agent(
+                agent=self,
```
Comment (jma) [inline rero_mef/agents/api.py:73]: It does not makes sense, remove class_method to `get_viaf_by_agent`.

File: rero_mef/agents/api.py [hunk 2]
```diff
@@ -72,31 +65,36 @@ def create_or_update_mef_viaf_record(self, dbcommit=False, reindex=False,
...
+                online=online
             )
-        else:
-            if mef_record:
+
+            ref_string = AgentMefRecord.build_ref_string(
+                agent=self.agent,
```
Comment (jma) [inline rero_mef/agents/api.py:78]: Same as previous.

Comment (jma) [reply rero_mef/agents/api.py:78]: Please add comments for each cases.

File: rero_mef/agents/api.py [hunk 3]
```diff
@@ -108,11 +106,20 @@ def create_or_update_mef_viaf_record(self, dbcommit=False, reindex=False,
...
                 )
-        if reindex:
-            AgentMefRecord.update_indexes()
-        return mef_record, mef_action, viaf_record, got_online
+            if reindex:
+                AgentMefRecord.update_indexes()
+            return mef_record, mef_action, viaf_record, got_online
+
+    @classmethod
+    def update_indexes(cls):
```
Comment (jma) [inline rero_mef/agents/api.py:116]: Why not rename update_indexes to flush_indexes?

File: rero_mef/api.py [hunk 4]
```diff
@@ -126,6 +126,9 @@ def update_test_md5(self, data, dbcommit=False, reindex=False):
         """
         data = deepcopy(data)
         data = add_md5(data)
+        print('--->')
```
Comment (jma) [inline rero_mef/api.py:129]: Can be removed, do not hesitate to do your own review.

File: rero_mef/api.py [hunk 5]
```diff
@@ -390,6 +393,11 @@ def get_metadata_identifier_names(cls):
         identifier = cls.provider.pid_identifier
         return metadata, identifier
 
+    @property
+    def deleted(self):
```
Comment (jma) [inline rero_mef/api.py:397]: Seems useless for me.

File: rero_mef/api_mef.py [hunk 6]
```diff
@@ -84,11 +84,9 @@ def get_all_pids_without_agents_and_viaf(cls):
 
         :returns: Generator of MEF pids without agent links and without VIAF.
         """
-        query = cls.search() \
-            .filter('bool', must_not=[Q('exists', field="viaf_pid")])
-        for pid_type in current_app.config.get(cls.mef_type, []):
-            query = query \
-                .filter('bool', must_not=[Q('exists', field=pid_type)])
+        must_not = [Q('exists', field="viaf_pid")]
```
Comment (jma) [inline rero_mef/api_mef.py:87]: You can use boolean operator i.e. `Q() | Q() & Q()`.

File: rero_mef/concepts/api.py [hunk 7]
```diff
@@ -100,25 +102,25 @@ def create_or_update_mef_viaf_record(self, dbcommit=False, reindex=False,
...
-        # if viaf_record:
-        #     mef_data['viaf_pid'] = viaf_record.pid
-        #     if not mef_record:
-        #         mef_record = ConceptMefRecord.get_mef_by_viaf_pid(
-        #             viaf_record.pid)
-        if self.deleted:
-            mef_record, mef_action = self.delete_from_mef(
-                dbcommit=dbcommit,
-                reindex=reindex
+        if not isinstance(self, ConceptMefRecord):
```
Comment (jma) [inline rero_mef/concepts/api.py:105]: Same as previous.

File: rero_mef/concepts/api.py [hunk 8]
```diff
@@ -100,25 +102,25 @@ def create_or_update_mef_viaf_record(self, dbcommit=False, reindex=False,
...
+            ref_string = ConceptMefRecord.build_ref_string(
+                concept=self.concept,
```
Comment (jma) [inline rero_mef/concepts/api.py:107]: same as previous.

File: rero_mef/concepts/api.py [hunk 9]
```diff
@@ -100,25 +102,25 @@ def create_or_update_mef_viaf_record(self, dbcommit=False, reindex=False,
...
+                concept_pid=self.pid
             )
-        else:
-            if mef_record:
+            mef_data = {self.concept: {'$ref': ref_string}}
+            mef_record = ConceptMefRecord.get_mef_by_entity_pid(
+                self.pid, self.name)
+            # if viaf_record:
```
Comment (jma) [inline rero_mef/concepts/api.py:113]: Please remove this.

File: run-tests.sh [hunk 10]
```diff
@@ -69,14 +69,12 @@ set -e
...
 # | celery                     | 5.1.2     | <5.2.2                   | 43738    |
-# | kombu                      | 5.1.0     | <5.2.1                   | 42497    |
 # | flask-security             | 3.0.0     | <3.1.0                   | 45183    |
 # | flask-security             | 3.0.0     | >0                       | 44501    |
 # | wtforms                    | 2.3.3     | <3.0.0a1                 | 42852    |
-# | flask-caching              | 1.10.1    | <=1.10.1                 | 40459    |
 # | sqlalchemy-utils           | 0.35.0    | >=0.27.0                 | 42194    |
 # +==============================================================================+
-safety check -i 47833 -i 42498 -i 43738 -i 42497 -i 45183 -i 44501 -i 42852 -i 40459 -i 42194
+safety check -i 47833 -i 42498 -i 43738 -i 45183 -i 44501 -i 42852 -i 42194
```
Comment (jma) [inline run-tests.sh:77]: Cool

File: tests/unit/test_api.py [hunk 11]
```diff
@@ -17,7 +17,10 @@
 
 """Test api."""
 
+from invenio_search import current_search_client
```
Comment (jma) [inline tests/unit/test_api.py:20]: Tests seems small compare to the changes...

**Review by jma — CHANGES_REQUESTED** [review]

File: tests/unit/test_api.py (hunk 11 above)
Comment (rerowep) [reply tests/unit/test_api.py:20]: Generally more test must be added.

**Pull request merged**